ADDED Option to separate numbers in snake Case

aayush123 opened the pull request:
This PR adds an option to separate numbers when using snake case, resolves #50 

Eg:

Earlier, `snakeCase('AddressLine1')` would give `'address_line1'`.

With new option, `snakeCase('AddressLine1', { separateNumbers: true })` would return `'address_line_1'`

blakeembrey commented:
Do you want to just add an option here: https://github.com/blakeembrey/change-case/blob/c24a0e18044bf77d9e824e93f04f7eb4e2151230/packages/no-case/src/index.ts#L20-L25?

This way would solve the issue for all "cases", since it's more in the interpretation of the string than the formatting side.

aayush123 commented:
If we do that, the regex changes that are done in snake case package would also be moved into no-case so that it can handle numbers there itself.
Would that be fine?

aayush123 commented:
@blakeembrey I have moved the option from snake-case to no-case.
Have tested the new option in no-case, dot-case, param-case, pascal-case & camel-case. Should I further test all packages for the effects the new option has?
Also, do note that the new test cases will pass only when we use the local no-case and in turn, local dot-case etc packages as dependencies. This should be resolved if and when we merge the changes in the repository.

blakeembrey commented on packages/no-case/src/index.ts:
```diff
@@ -17,12 +19,16 @@ const DEFAULT_STRIP_REGEXP = /[^A-Z0-9]+/gi;
 export function noCase(input: string, options: Options = {}) {
-  const {
-    splitRegexp = DEFAULT_SPLIT_REGEXP,
+  let {
```
Prefer to keep `const`, you can either destructure it later to re-use earlier variables or just make it a new constant in a separate statement. No reason to use `let` here.

aayush123 replied:
Yeah makes sense. Following the latter conversation, I'll resolve this by appending the new regexes to the regex list instead of making a new list.

blakeembrey commented on packages/no-case/src/index.ts:
```diff
@@ -5,10 +5,12 @@ export interface Options {
 
 // Support camel case ("camelCase" -> "camel Case" and "CAMELCase" -> "CAMEL Case").
 const DEFAULT_SPLIT_REGEXP = [/([a-z0-9])([A-Z])/g, /([A-Z])([A-Z][a-z])/g];
+const SEPARATE_NUMBERS_SPLIT_REGEXP = [/([a-z0-9])([A-Z])/g, /([A-Z])([A-Z][a-z])/g, /([(0-9)])([A-Za-z])/g, /([A-Za-z])([0-9])/g,];
```
What about appending the two additions you want to the regex list when the option is specified? Also, can we get some more test coverage, specifically around the final regex (number then letter) and combinations of things like `Foo12019Bar` or `V1Test`.

blakeembrey replied:
```suggestion
const SEPARATE_NUMBERS_SPLIT_REGEXP = [/([a-z0-9])([A-Z])/g, /([A-Z])([A-Z][a-z])/g, /([0-9])([A-Za-z])/g, /([A-Za-z])([0-9])/g,];
```

Typo in the number group.

aayush123 commented:
> What about appending the two additions you want to the regex list when the option is specified?

Appending wouldn't particularly work cause I guess we should add the 2 new regexps only when user didn't specify their own `splitRegexp`.
So I resorted to creating a new array using the default one and adding the new options to it.
Afterwards, based on whether or not the user provided `splitRegexp` and `separateNumbers` options, either new or default regexp is used.
Also updated code to de-structure `splitRegexp` separately to use `let` only for this.

> Also, can we get some more test coverage, specifically around the final regex (number then letter) and combinations of things like Foo12019Bar or V1Test.

Added test cases in various packages.
The new option does provide results that may not particularly be useful in `camelCase` and `PascalCase` (the two that I tested). These two in particular add an underscore before numbers when `separateNumbers` is `true`. I wonder if it'll be useful in any scenario.
For eg.
```typescript
pascalCase("Foo12019Bar", { separateNumbers: true });    //  "Foo_12019Bar"
pascalCase("aNumber2in", { separateNumbers: true });     //  "ANumber_2In"
pascalCase("V1Test", { separateNumbers: true });         //  "V_1Test"
```

MichaelGregory commented:
Is there still interest in merging this feature? I welcome it.
